File: collabcanvas/src/types/extraction.ts
/**
 * ExtractionResult types for Epic 3 PR2 (CAD parsing).
 *
 * These mirror the CADData / SpaceModel / SpatialRelationships
 * sections in `docs/clarification-output-schema.md` and are aligned
 * with the Python TypedDicts in `functions_py/types/extraction.py`.
 */

export type FileType = 'dwg' | 'dxf' | 'pdf' | 'png' | 'jpg';
export type ExtractionMethod = 'ezdxf' | 'vision';

export interface Dimensions {
  length: number;
  width: number;
  height?: number;
}

export interface Room {
  id: string;
  name: string;
  type: string; // "kitchen", "bathroom", etc.
  sqft: number;
  dimensions: Dimensions;
  confidence: number;
  needsVerification: boolean;
}

export type WallType = 'interior' | 'exterior' | 'load_bearing' | 'partition';

export interface Wall {
  id: string;
  length: number;
  height?: number;
  thickness?: number;
  type: WallType;
  material?: string;
  connectsRooms: string[];
  adjacentWalls: string[];
  confidence: number;
}

export type OpeningType = 'door' | 'window' | 'archway' | 'pass_through';

export type OpeningSide = 'left' | 'right' | 'center';

export type OpeningSwing =
  | 'in'
  | 'out'
  | 'left'
  | 'right'
  | 'sliding'
  | 'pocket';

export interface OpeningPosition {
  distanceFromCorner: number;
  side: OpeningSide;
}

export interface Opening {
  id: string;
  type: OpeningType;
  width: number;
  height: number;
  inWall: string;
  connectsRooms: string[];
  position: OpeningPosition;
  swing?: OpeningSwing;
  confidence: number;
}

export type LengthUnits = 'feet' | 'inches' | 'meters';

export interface BoundingBox {
  length: number;
  width: number;
  height: number;
  units: LengthUnits;
}

export interface Scale {
  detected: boolean;
  ratio?: number;
  units: LengthUnits;
}

export interface SpaceModel {
  totalSqft: number;
  boundingBox: BoundingBox;
  scale: Scale;
  rooms: Room[];
  walls: Wall[];
  openings: Opening[];
}

export type RoomConnectionType = 'door' | 'archway' | 'open' | 'window';

export interface RoomAdjacency {
  room1: string;
  room2: string;
  connection: RoomConnectionType;
  openingId?: string;
}

export interface EntryPoint {
  openingId: string;
  fromSpace: string; // "hallway", "exterior", "garage", etc.
  isPrimary: boolean;
}

export interface SpatialRelationships {
  layoutNarrative: string;
  roomAdjacencies: RoomAdjacency[];
  entryPoints: EntryPoint[];
}

export interface ExtractionResult {
  // File info
  fileUrl: string;
  fileType: FileType;
  extractionMethod: ExtractionMethod;
  extractionConfidence: number; // 0-1 overall confidence

  // Physical space & relationships
  spaceModel: SpaceModel;
  spatialRelationships: SpatialRelationships;

  // Raw extraction payload from parser / model (for debugging)
  rawExtraction?: Record<string, unknown>;
}

File: functions_py/tc_types/extraction.py
"""
Non-conflicting copy of CAD extraction TypedDicts.

This file mirrors the content of the former `functions_py/types/extraction.py`
so we can avoid shadowing Python's standard library `types` module.
"""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional, TypedDict


FileType = Literal["dwg", "dxf", "pdf", "png", "jpg"]
ExtractionMethod = Literal["ezdxf", "vision"]


class Dimensions(TypedDict, total=False):
    """Room dimensions."""

    length: float
    width: float
    height: Optional[float]


class Room(TypedDict, total=False):
    """Simplified room representation from CAD."""

    id: str
    name: str
    type: str  # e.g. "kitchen", "bathroom"
    sqft: float
    dimensions: Dimensions
    confidence: float
    needsVerification: bool


class Wall(TypedDict, total=False):
    """Wall segment between rooms."""

    id: str
    length: float
    height: Optional[float]
    thickness: Optional[float]
    type: Literal["interior", "exterior", "load_bearing", "partition"]
    material: Optional[str]
    connectsRooms: List[str]
    adjacentWalls: List[str]
    confidence: float


class OpeningPosition(TypedDict, total=False):
    """Relative position of an opening in a wall."""

    distanceFromCorner: float
    side: Literal["left", "right", "center"]


class Opening(TypedDict, total=False):
    """Door / window / opening in a wall."""

    id: str
    type: Literal["door", "window", "archway", "pass_through"]
    width: float
    height: float
    inWall: str
    connectsRooms: List[str]
    position: OpeningPosition
    swing: Literal["in", "out", "left", "right", "sliding", "pocket"]
    confidence: float


class BoundingBox(TypedDict, total=False):
    """Overall bounding box of the project space."""

    length: float
    width: float
    height: float
    units: Literal["feet", "inches", "meters"]


class Scale(TypedDict, total=False):
    """Scale information for the CAD drawing."""

    detected: bool
    ratio: Optional[float]
    units: Literal["feet", "inches", "meters"]


class SpaceModel(TypedDict, total=False):
    """
    Physical space model derived from CAD.

    This aligns with `SpaceModel` in the ClarificationOutput schema:
    - totalSqft
    - boundingBox
    - scale
    - rooms
    - walls
    - openings
    """

    totalSqft: float
    boundingBox: BoundingBox
    scale: Scale
    rooms: List[Room]
    walls: List[Wall]
    openings: List[Opening]


class RoomAdjacency(TypedDict, total=False):
    """Adjacency relationship between two rooms."""

    room1: str
    room2: str
    connection: Literal["door", "archway", "open", "window"]
    openingId: Optional[str]


class EntryPoint(TypedDict, total=False):
    """Entry points into the project space."""

    openingId: str
    fromSpace: str  # e.g. "hallway", "exterior", "garage"
    isPrimary: bool


class SpatialRelationships(TypedDict, total=False):
    """
    Narrative + structural relationships between rooms.

    Mirrors `SpatialRelationships` in the ClarificationOutput spec.
    """

    layoutNarrative: str
    roomAdjacencies: List[RoomAdjacency]
    entryPoints: List[EntryPoint]


class ExtractionResult(TypedDict, total=False):
    """
    Unified extraction result for both ezdxf and GPT-4o Vision paths.

    This is intentionally shaped like `CADData` so that later PRs can
    drop it directly into ClarificationOutput.cadData with minimal
    transformation.
    """

    # File info
    fileUrl: str
    fileType: FileType
    extractionMethod: ExtractionMethod
    extractionConfidence: float  # 0-1 overall confidence

    # Physical space & relationships
    spaceModel: SpaceModel
    spatialRelationships: SpatialRelationships

    # Raw extraction payload from parser / model (for debugging)
    rawExtraction: Dict[str, Any]


__all__ = [
    "FileType",
    "ExtractionMethod",
    "Dimensions",
    "Room",
    "Wall",
    "OpeningPosition",
    "Opening",
    "BoundingBox",
    "Scale",
    "SpaceModel",
    "RoomAdjacency",
    "EntryPoint",
    "SpatialRelationships",
    "ExtractionResult",
]



